[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #168129

(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

Resolves #167682

fduwjj committed:
[Symm mem] Make nccl backend full fleged

[ghstack-poisoned]

fduwjj committed:
Update on "[Symm mem] Make nccl backend full fleged"




[ghstack-poisoned]

fduwjj committed:
Update on "[Symm mem] Make nccl backend full fleged"




[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/symm_mem/NCCLSymmetricMemory.cu:
```diff
@@ -43,13 +44,15 @@ class NCCLSymmetricMemory : public SymmetricMemory {
       std::shared_ptr<NCCLAllocation> allocation,
       const std::string& group_name,
```
somehow the signature of base class is a reference. so I found that calling std::move of a value here will lead to error. (Later program cannot reference to the group_name)

kwen2501 commented on test/distributed/test_nccl.py:
```diff
@@ -250,6 +250,101 @@ def foo():
+        if self.rank == 0:
+            torch.ops.symm_mem.nccl_get(tensor, 1)
+            # TODO: remove after we have wait_signal
+            c10d.barrier()
```
Should the barrier be before get?

kwen2501 replied:
rendezvous does not provide GPU-side barrier guarantee.

kwen2501 commented on torch/csrc/distributed/c10d/symm_mem/nccl_extension.cu:
```diff
@@ -0,0 +1,215 @@
+#include <dlfcn.h>
```
What is this header for?

fduwjj replied:
let me removed it.

kwen2501 commented on torch/csrc/distributed/c10d/symm_mem/nccl_extension.cu:
```diff
@@ -0,0 +1,215 @@
+
+    for (size_t i = tid; i < nelems; i += stride) {
+        dst[i] = src[i];
+    }
```
Same

kwen2501 reviewed (approved):
looks good. 
Can you please have a look at the comments? Thanks.

fduwjj commented on test/distributed/test_nccl.py:
```diff
@@ -250,6 +250,101 @@ def foo():
+            torch.ops.symm_mem.nccl_wait_for_signal(tensor, signal_val, 1)
+            torch.testing.assert_close(
+                tensor, torch.zeros(numel, dtype=dtype, device=self.device)
+            )
```
sure.

ngimel commented on torch/csrc/distributed/c10d/symm_mem/nccl_extension.cu:
```diff
@@ -0,0 +1,215 @@
+
+        while (true) {
+            unsigned long long val = *sig_ptr;
+            if (val >= static_cast<unsigned long long>(target_signal_value)) break;
```
`>=` is typically used to avoid deadlocks, if peer increments your signal pad twice before you had a chance to read it it would be a deadlock otherwise. Of course, it requires incrementing twice to be safe, and not mean that you've missed something and never see it again. Anyways, nccl is using >= for its things

fduwjj commented on torch/csrc/distributed/c10d/symm_mem/NCCLSymmetricMemory.cu:
```diff
@@ -35,26 +37,89 @@ struct NCCLAllocation {
+  int tid = blockIdx.x * blockDim.x + threadIdx.x;
+  int stride = blockDim.x * gridDim.x;
+  for (int peer = tid; peer < world_size; peer += stride) {
+      table[peer] = ncclGetLsaPointer(handle, offset, peer);
```
No... At least I didn't find one. Now you are in NV and comm department, you can tell them to add one. lol

fduwjj commented on torch/csrc/distributed/c10d/symm_mem/SymmetricMemory.cpp:
```diff
@@ -503,6 +503,11 @@ TORCH_LIBRARY_FRAGMENT(symm_mem, m) {
+  m.def("nccl_get(Tensor(a!) tensor, int peer) -> ()");
+  m.def("nccl_wait_for_signal(Tensor sigpad, int signal, int peer) -> ()");
+  m.def(
+      "nccl_put_with_signal(Tensor(a) tensor, int signal, int peer) -> ()");
```
well the signal pad and tensor are actually in the same tensor. But I guess I can pass in the same tensors twice.

fduwjj commented on torch/csrc/distributed/c10d/symm_mem/nccl_extension.cu:
```diff
@@ -0,0 +1,215 @@
+
+    for (size_t i = tid; i < nelems; i += stride) {
+        dst[i] = src[i];
+    }
```
Sure.

fduwjj commented on torch/csrc/distributed/c10d/symm_mem/nccl_extension.cu:
```diff
@@ -0,0 +1,215 @@
+
+    // Calculate remote dst pointer
+    T* dst = reinterpret_cast<T*>(
+        get_remote_ptr(buffer, dst_peer, dst_byte_offset));
```
which one are you talking about?

fduwjj commented on torch/csrc/distributed/c10d/symm_mem/nccl_extension.cu:
```diff
@@ -0,0 +1,215 @@
+
+__global__ void nccl_wait_for_signal_table_kernel(
+    void**  signal_pad,
+    int  cur_rank,
```
hmm I was thinking about we only set signals for the rank where the data will be put. Do we also want to set it for the ranks which puts the data?

fduwjj commented on torch/csrc/distributed/c10d/symm_mem/nccl_extension.cu:
```diff
@@ -0,0 +1,215 @@
+    return base + byte_offset;
+}
+
+template <typename T>
```
hmm I just see cuda kernel is doing the similar thing for tensors as well. And when I try without template, the numerical does not match..

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

[ghstack-poisoned]

fduwjj commented on torch/csrc/distributed/c10d/symm_mem/nccl_extension.cu:
```diff
@@ -0,0 +1,239 @@
+    __syncthreads();
+
+    // 2) system fence + signal
+    if (tid == 0) {
```
OK I see what you mean here and I add a global counter across blocks to ensure the signal is only set after all blocks finished writing.

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

[ghstack-poisoned]

fduwjj committed:
Update

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

Resolves #167682

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

Resolves #167682

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

Resolves #167682

[ghstack-poisoned]

kwen2501 commented on .ci/pytorch/test.sh:
```diff
@@ -373,7 +373,7 @@ test_h100_symm_mem() {
   time python test/run_test.py --include distributed/test_nvshmem.py $PYTHON_TEST_EXTRA_OPTION --upload-artifacts-while-running
   time python test/run_test.py --include distributed/test_nvshmem_triton.py $PYTHON_TEST_EXTRA_OPTION --upload-artifacts-while-running
-  time python test/run_test.py --include distributed/test_nccl.py $PYTHON_TEST_EXTRA_OPTION --upload-artifacts-while-running
+  time python test/run_test.py --include distributed/test_nccl.py -k NCCLSymmetricMemoryTest $PYTHON_TEST_EXTRA_OPTION --upload-artifacts-while-running
```
nit: consider moving it out as a separate file? There could be more TestCases coming

fduwjj commented on .ci/docker/common/install_inductor_benchmark_deps.sh:
```diff
@@ -55,3 +55,8 @@ install_timm
+  conda_run pip uninstall -y nvidia-nccl-cu13
+else
+  conda_run pip uninstall -y nvidia-nccl-cu12
+fi
```
yes because we look into the docker logs and found that nccl was installed via pip which overrides the default NCCL.

fduwjj replied:
right but uninstall nvidia-nccl-* does not work

fduwjj commented:
@pytorchbot merge

kwen2501 commented on torch/csrc/distributed/c10d/symm_mem/NCCLSymmetricMemory.cu:
```diff
@@ -2,11 +2,13 @@
 #include <torch/csrc/cuda/nccl.h>
 
-#if NCCL_VERSION_CODE >= NCCL_VERSION(2, 27, 1)
+#if NCCL_VERSION_CODE >= NCCL_VERSION(2, 28, 9)
```
Usually features are the same within a minor version (2.28), not sensitive to patch version (2.28.9). We can use (2, 28, 0) here.

fduwjj commented on torch/csrc/distributed/c10d/symm_mem/NCCLSymmetricMemory.cu:
```diff
@@ -238,8 +326,19 @@ class NCCLSymmetricMemoryAllocator : public SymmetricMemoryAllocator {
+    // See https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/device.html#nccldevcommrequirements
+    int nCTAs = 16;
+    reqs.lsaBarrierCount = nCTAs;
+    C10D_NCCL_CHECK(ncclDevCommCreate(comm, &reqs, &devComm), "ncclDevCommCreate failed");
```
Maybe we should cache by comm?

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

Resolves #167682

[ghstack-poisoned]

albanD commented:
@pytorchbot  revert -c weird

The inductor test fails when compiling nccl code.
This is super weird that it is flaky compilation failure though.

kwen2501 commented on torch/csrc/distributed/c10d/symm_mem/nccl_extension.cuh:
```diff
@@ -0,0 +1,26 @@
+#include <torch/csrc/cuda/nccl.h>
+#if NCCL_VERSION_CODE >= NCCL_VERSION(2, 28, 0)
+#define NCCL_HAS_SYMMEM_SUPPORT
+#endif
```
This will entirely disable NCCL SymmMem for builds with NCCL 2.27? torch 2.10 CU12 build still uses 2.27.

fduwjj replied:
I really don't like this idea of multiple macros, but yes we can do that.

tinglvv commented:
Suggesting to run the ciflow/binaries test as well as this PR added nccl_extension.cu into the build.

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

Resolves #167682

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

Resolves #167682

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d][Sym mem] Make nccl backend full fledged with nccl 2.28.9-1"


(This PR will be rebased on https://github.com/pytorch/pytorch/pull/166174) (There are other PR which updates NCCL version: https://github.com/pytorch/pytorch/pull/168091)

We did the following thing:
1. To add exchange of buffer ptr and signal pad ptr via NCCL device API introduced in nccl 2.28.
2. With #1, we showed that the symmem from nccl backend works with existing one_shot_all_reduce kernel (Add a UT for it)
3. Add a simple put, put with signal, wait for signal and get. So that symmem's one side API works.
4. Show that symmem from nccl backend works with traditional c10d collective as well in UT.
5. Stored DevComm inside symmetric memory so that users can access to it for customized kernel. 

Resolves #167682

[ghstack-poisoned]
